### DISCUSSION TITLE
Generic Type Parsing Rejection Function

### DISCUSSION THREAD

**Discussion by Mortuie** [colinhacks/zod#701] [Q&A]
So i'm writing a REST api for a pet project (i know i do love graphql but just brushing up and learning a bit about REST)

And because of very little type safety I need to validate the params that come into my endpoints

This ends up looking a lot like this:

```js

  const temp = zz.safeParse(objectToParse);

  if (!temp.success) {
    res.status(400).json(INVALID_PARAMETERS);
  }
```

However rather than writing this out every time I was wondering if someone had any idea how to generic this up because I don't something like:

`const p = (zz, objectToParse, res) => {`
Where zz is the zod object, and res is the express Response, so essentially if objectToParse conforms to zz I want to return the parsedVersion if not I want to res.send(400).json(INVALID PARAMS);

Anyone got any idea how to do this?

Thanks in advance!

**Comment by JacobWeisenburger** [comment]
I'm pretty sure you'll want to use [express middleware](https://expressjs.com/en/guide/using-middleware.html)

```ts
import { z } from 'zod'
import express, { RequestHandler } from 'express'
const app = express()

function validateParams<Schema extends z.AnyZodObject> ( schema: Schema ): RequestHandler {
    return ( req, res, next ) => { // express middleware function
        const parseResult = schema.safeParse( req.params )

        if ( parseResult.success ) {
            next()
        } else {
            console.error( parseResult.error.message )
            res.status( 400 ).json( INVALID_PARAMETERS )
        }
    }
}

// you can use middleware in any route
app.use(
    '/user/:id',
    validateParams(
        z.object( {
            id: z.string()
        } )
    ),
    ( req, res ) => res.send( 'Success!' )
)
```
